rust/src/editor.rs

```diff
@@ -97,9 +99,100 @@ impl Editor {
+            MoveRight => async(state.move_right(0)),
+            MoveRightAndModifySelection => async(state.move_right(FLAG_SELECT)),
+            MoveToBeginningOfParagraph => async(state.cursor_start()),
+            MoveToEndOfParagraph => async(state.cursor_end()),
+            MoveToLeftEndOfLine => async(state.move_to_left_end_of_line(0)),
+            MoveToLeftEndOfLineAndModifySelection => async(state.move_to_left_end_of_line(FLAG_SELECT)),
+            MoveToRightEndOfLine => async(state.move_to_right_end_of_line(0)),
+            MoveToRightEndOfLineAndModifySelection => async(state.move_to_right_end_of_line(FLAG_SELECT)),
+            MoveToBeginningOfDocument => async(state.move_to_beginning_of_document(0)),
+            MoveToBeginningOfDocumentAndModifySelection => async(state.move_to_beginning_of_document(FLAG_SELECT)),
+            MoveToEndOfDocument => async(state.move_to_end_of_document(0)),
+            MoveToEndOfDocumentAndModifySelection => async(state.move_to_end_of_document(FLAG_SELECT)),
+            ScrollPageUp => async(state.scroll_page_up(0)),
+            PageUpAndModifySelection => async(state.scroll_page_up(FLAG_SELECT)),
+            ScrollPageDown => async(state.scroll_page_down(0)),
+            PageDownAndModifySelection => {
+                async(state.scroll_page_down(FLAG_SELECT))
+            }
+            Open { file_path } => async(state.do_open(file_path)),
+            Save { file_path } => async(state.do_save(file_path)),
+            Scroll { first, last } => async(state.do_scroll(first, last)),
+            Yank => async(state.yank()),
+            Transpose => async(state.do_transpose()),
+            Click { line, column, flags, click_count } => {
+                async(state.do_click(line, column, flags, click_count))
+            }
+            Drag { line, column, flags } => async(state.do_drag(line, column, flags)),
+            Undo => async(state.do_undo()),
+            Redo => async(state.do_redo()),
+            Cut => Some(state.do_cut()),
+            Copy => Some(state.do_copy()),
+            DebugRewrap => async(state.debug_rewrap()),
+            DebugTestFgSpans => async(state.debug_test_fg_spans()),
+            DebugRunPlugin => async(self.debug_run_plugin()),
+        };
+
+        // TODO: could defer this until input quiesces - will this help?
+        state.commit_delta();
+        state.render();
+        state.last_edit_type = state.this_edit_type;
+        state.gc_undos();
+        result
+    }
+
+    fn debug_run_plugin(&self) {
+        print_err!("running plugin");
+        start_plugin(self.clone());
+    }
+
+    pub fn on_plugin_connect(&self, plugin_ref: PluginRef) {
+        let mut state = self.0.lock().unwrap();
+        plugin_ref.ping_from_editor(state.plugin_buf_size());
+        state.plugins.push(plugin_ref);
+    }
+
+    pub fn downgrade(&self) -> Weak<Mutex<EditorState>> {
+        Arc::downgrade(&self.0)
```

Comment:
With this part, I am currently least happy with, because it creates the requirement of making the `EditorState` `pub`lic